### ISSUE TITLE
Current controls for the model viewer aren’t trackpad friendly

### ISSUE THREAD

**Issue by cvlnomen** [Kizari/Flagrum#20]
The current controls used for the model viewer make it impossible for trackpad users to rotate and pan a 3D model. The reason for this is that a trackpad cannot hold down the middle mouse button.

Possible solution: add some bindings so the model viewer can work in a similar way like trackpad friendly model viewers.

**Comment by Kizari** [comment]
Thanks for this. The current keybindings were setup to mirror Blender since the Flagrum add-on is for that software and is what users are accustomed to. However, I see no harm in a second set of bindings for those that need them.

Ideally these would be configurable by the user (i.e. bind the Gesture to a property on MainViewModel and hook that up to WpfService to be called from some UI in SettingsPage in Flagrum.Web). However, I don't have the time for this myself, so unless you're willing to do that, this can be merged as a compromise.

**Comment by cvlnomen** [comment]
Sounds like a fun challenge, so I'll give it a try when I have some time off. 
In what way should the user be able to adjust his keybindings? Should he be given the option to select a preset, or should it be possible to set each command separately?

**Comment by Kizari** [comment]
It'd be quite involved, let me explain.

`Flagrum.Web/Features/AssetExplorer/Previews/ModelPreviewV2.razor` controls the web-client side of the 3D viewer. The viewer is overlaid on the bounds of "viewportContainer."

```
<div id="viewportContainer" class="h-full bg-grey-700 row">
    @if (IsLoading)
    {
        <div class="flex flex-col items-center w-full">
            <div class="continuous-3"></div>
            <span class="block text-accent1-200 font-display font-bold mt-4 mr-2">Loading Model</span>
        </div>
    }
</div>
```

I was thinking of adding a thin strip above and below this (same background colour) and using the top strip to display the key bindings, and the bottom to have something like a pause and play for animations and a slider for the frames once animation support is added later. In terms of configuration for the 3D viewer, the right end of the top strip would probably be a good place for a settings button, which could pop up an `AutoSizeModal` component with the 3D viewer settings. You also need to hide the 3D viewer when the modal is opened and show it again when it's closed (see `Flagrum.Web/Features/AssetExplorer/Export/ExportWithDependenciesModal.razor` for an example) as otherwise it will overlap the modal and softlock the application.

I would personally then have the option to choose a binding for the rotate and pan as separate dropdowns. You may need to research into that `Gesture` field to see which options are a possibility. Upon closing this modal, I would persist the settings to the database as a `StatePair` and then call `WpfService` to update the bound gestures on `MainViewModel`. Of course, you would also need to load these from the DB when the application first starts, such as in the `MainViewModel` constructor. There are examples of all these things throughout the application if you need some code to learn from.

I appreciate that this project is a little unorthodox due to its hybrid WPF/Blazor nature, so feel free to ask any questions.

**Comment by cvlnomen** [comment]
Hey I have two questions:
1. What's your reason for saving the settings as a StatePair? Can't I just use Context.SetString?
2. The first time a user will run Flagrum with the new settings menu, there won't be a value for the gesture settings in the DB. In which class should the default value be created/set when the user runs the application for the first time?

**Comment by Kizari** [comment]
1. This is the same thing. Context.SetString is just a method I created to simplify the process of managing state pairs.
2. When you call Context.GetString, if there is no value set, it will simply return null. You can do something like the following:

```
KeybindRotate = Context.GetString(StatePairs.KeybindRotate3DViewer) ?? "MiddleMouse";
```

**Comment by Kizari** [comment]
This was automatically closed by merging the PR.

Reopening as there is still some work to do by the maintainer to complete this issue, namely:

> @Kizari to update the UI to overlay the 3D view to resolve the clipping issue, make some minor cosmetic improvements as requested, and add the Japanese localisation.

**Comment by Kizari** [comment]
This has now been completed with commit [72674cbf94a936dda2cb23825a462cc6c855bbd0](https://github.com/Kizari/Flagrum/commit/72674cbf94a936dda2cb23825a462cc6c855bbd0).  
This will be released as part of Flagrum 1.2.7.
